File: bindings/http/src/test/java/org/openecard/control/binding/http/TestClient.java
/****************************************************************************
 * Copyright (C) 2012-2019 ecsec GmbH.
 * All rights reserved.
 * Contact: ecsec GmbH (info@ecsec.de)
 *
 * This file is part of the Open eCard App.
 *
 * GNU General Public License Usage
 * This file may be used under the terms of the GNU General Public
 * License version 3.0 as published by the Free Software Foundation
 * and appearing in the file LICENSE.GPL included in the packaging of
 * this file. Please review the following information to ensure the
 * GNU General Public License version 3.0 requirements will be met:
 * http://www.gnu.org/copyleft/gpl.html.
 *
 * Other Usage
 * Alternatively, this file may be used in accordance with the terms
 * and conditions contained in a signed written agreement between
 * you and ecsec GmbH.
 *
 ***************************************************************************/

package org.openecard.control.binding.http;

import iso.std.iso_iec._24727.tech.schema.CardInfoType;
import iso.std.iso_iec._24727.tech.schema.ConnectionHandleType;
import iso.std.iso_iec._24727.tech.schema.EstablishContext;
import iso.std.iso_iec._24727.tech.schema.EstablishContextResponse;
import java.io.InputStream;
import org.openecard.addon.AddonManager;
import org.openecard.common.ClientEnv;
import org.openecard.common.event.EventDispatcherImpl;
import org.openecard.common.interfaces.CIFProvider;
import org.openecard.common.interfaces.EventDispatcher;
import org.openecard.gui.swing.SwingDialogWrapper;
import org.openecard.gui.swing.SwingUserConsent;
import org.openecard.ifd.scio.IFD;
import org.openecard.management.TinyManagement;
import org.openecard.recognition.CardRecognitionImpl;
import org.openecard.sal.TinySAL;
import org.openecard.transport.dispatcher.MessageDispatcher;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


/**
 * Implements a TestClient to test the HttpBinding.
 *
 * @author Moritz Horsch
 * @author Dirk Petrautzki
 */
public final class TestClient {

    private static final Logger LOG = LoggerFactory.getLogger(TestClient.class);

    // Service Access Layer (SAL)
    private TinySAL sal;

    public TestClient() {
	try {
	    setup();
	} catch (Exception e) {
	    LOG.error(e.getMessage(), e);
	}
    }

    private void setup() throws Exception {
	// Set up client environment
	ClientEnv env = new ClientEnv();

	// Set up the IFD
	IFD ifd = new IFD();
	env.setIFD(ifd);

	// Set up Management
	TinyManagement management = new TinyManagement(env);
	env.setManagement(management);

	// Set up the Dispatcher
	MessageDispatcher dispatcher = new MessageDispatcher(env);
	env.setDispatcher(dispatcher);

	// Perform an EstablishContext to get a ContextHandle
	EstablishContext establishContext = new EstablishContext();
	EstablishContextResponse establishContextResponse = ifd.establishContext(establishContext);

	byte[] contextHandle = ifd.establishContext(establishContext).getContextHandle();

	final CardRecognitionImpl recognition = new CardRecognitionImpl(env);
	env.setRecognition(recognition);

	env.setCIFProvider(new CIFProvider() {
	    @Override
	    public CardInfoType getCardInfo(ConnectionHandleType type, String cardType) {
		return recognition.getCardInfo(cardType);
	    }
	    @Override
	    public boolean needsRecognition(byte[] atr) {
		return true;
	    }
            @Override
            public CardInfoType getCardInfo(String cardType) throws RuntimeException {
                return recognition.getCardInfo(cardType);
            }
            @Override
            public InputStream getCardImage(String cardType) {
                return recognition.getCardImage(cardType);
            }
	});

	// Set up EventManager
	EventDispatcher ed = new EventDispatcherImpl();
	env.setEventDispatcher(ed);

	// Set up SALStateCallback
	// TODO: fix tests
//	cardStates = new CardStateMap();
//	SALStateCallback salCallback = new SALStateCallback(env, cardStates);
//	ed.add(salCallback);
//
//	// Set up SAL
//	sal = new TinySAL(env, cardStates);
//	env.setSAL(sal);
//
//	// Set up GUI
//	SwingUserConsent gui = new SwingUserConsent(new SwingDialogWrapper());
//	sal.setGUI(gui);
//	ifd.setGUI(gui);
//
//	// Initialize the EventManager
//	ed.start();
//
//	AddonManager manager = new AddonManager(env, gui, null);
//	sal.setAddonManager(manager);
//
//	HttpBinding binding = new HttpBinding(24727);
//	binding.setAddonManager(manager);
//	binding.start();
    }

}

File: bindings/http/src/jvmTest/java/org/openecard/control/binding/http/TestClient.kt
/****************************************************************************
 * Copyright (C) 2012-2019 ecsec GmbH.
 * All rights reserved.
 * Contact: ecsec GmbH (info@ecsec.de)
 *
 * This file is part of the Open eCard App.
 *
 * GNU General Public License Usage
 * This file may be used under the terms of the GNU General Public
 * License version 3.0 as published by the Free Software Foundation
 * and appearing in the file LICENSE.GPL included in the packaging of
 * this file. Please review the following information to ensure the
 * GNU General Public License version 3.0 requirements will be met:
 * http://www.gnu.org/copyleft/gpl.html.
 *
 * Other Usage
 * Alternatively, this file may be used in accordance with the terms
 * and conditions contained in a signed written agreement between
 * you and ecsec GmbH.
 *
 */
package org.openecard.control.binding.http

import io.github.oshai.kotlinlogging.KotlinLogging
import iso.std.iso_iec._24727.tech.schema.CardInfoType
import iso.std.iso_iec._24727.tech.schema.ConnectionHandleType
import iso.std.iso_iec._24727.tech.schema.EstablishContext
import org.openecard.common.ClientEnv
import org.openecard.common.event.EventDispatcherImpl
import org.openecard.common.interfaces.CIFProvider
import org.openecard.common.interfaces.EventDispatcher
import org.openecard.ifd.scio.IFD
import org.openecard.management.TinyManagement
import org.openecard.recognition.CardRecognitionImpl
import org.openecard.sal.TinySAL
import org.openecard.transport.dispatcher.MessageDispatcher
import java.io.InputStream

val LOG = KotlinLogging.logger {}

/**
 * Implements a TestClient to test the HttpBinding.
 *
 * @author Moritz Horsch
 * @author Dirk Petrautzki
 */
class TestClient {
	// Service Access Layer (SAL)
	private val sal: TinySAL? = null

	init {
		try {
			setup()
		} catch (e: Exception) {
			LOG.error(e) { "${e.message}" }
		}
	}

	@Throws(Exception::class)
	private fun setup() {
		// Set up client environment
		val env = ClientEnv()

		// Set up the IFD
		val ifd = IFD()
		env.ifd = ifd

		// Set up Management
		val management = TinyManagement(env)
		env.management = management

		// Set up the Dispatcher
		val dispatcher = MessageDispatcher(env)
		env.dispatcher = dispatcher

		// Perform an EstablishContext to get a ContextHandle
		val establishContext = EstablishContext()
		val establishContextResponse = ifd.establishContext(establishContext)

		val contextHandle = ifd.establishContext(establishContext).contextHandle

		val recognition = CardRecognitionImpl(env)
		env.recognition = recognition

		env.cifProvider =
			object : CIFProvider {
				override fun getCardInfo(
					type: ConnectionHandleType?,
					cardType: String,
				): CardInfoType? = recognition.getCardInfo(cardType)

				override fun needsRecognition(atr: ByteArray): Boolean = true

				@Throws(RuntimeException::class)
				override fun getCardInfo(cardType: String): CardInfoType? = recognition.getCardInfo(cardType)

				override fun getCardImage(cardType: String): InputStream? = recognition.getCardImage(cardType)
			}

		// Set up EventManager
		val ed: EventDispatcher = EventDispatcherImpl()
		env.eventDispatcher = ed

		// Set up SALStateCallback
		// TODO: fix tests
// 	cardStates = new CardStateMap();
// 	SALStateCallback salCallback = new SALStateCallback(env, cardStates);
// 	ed.add(salCallback);
//
// 	// Set up SAL
// 	sal = new TinySAL(env, cardStates);
// 	env.setSAL(sal);
//
// 	// Set up GUI
// 	SwingUserConsent gui = new SwingUserConsent(new SwingDialogWrapper());
// 	sal.setGUI(gui);
// 	ifd.setGUI(gui);
//
// 	// Initialize the EventManager
// 	ed.start();
//
// 	AddonManager manager = new AddonManager(env, gui, null);
// 	sal.setAddonManager(manager);
//
// 	HttpBinding binding = new HttpBinding(24727);
// 	binding.setAddonManager(manager);
// 	binding.start();
	}
}
